Pod Topology Spread blog post (KEP-3022 KEP-3094 KEP-3243) (#39777)
kubernetes/website: content/en/blog/_posts/2023-04-11-topology-spread-features.md

---
layout: blog
title: "Kubernetes 1.27: More fine-grained pod topology spread policies reached beta"
date: 2023-04-11
slug: fine-grained-pod-topology-spread-features-beta
evergreen: true
---

**Authors:** [Alex Wang](https://github.com/denkensk)(Shopee), [Kante Yin](https://github.com/kerthcet)(DaoCloud), [Kensei Nakada](https://github.com/sanposhiho)(Mercari)

In Kubernetes v1.19, [Pod Topology Spread Constraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) went to GA.
It is the feature to control how Pods are spread to each failure-domain (regions, zones, nodes etc).

As time passes, we've got further feedbacks from users,
and we're actively working on improving the Topology Spread via three KEPs from v1.25.
All of these features have reached beta in Kubernetes v1.27 and been enabled by default.

This blog post is going to introduce each feature and the usecase/issue behind them.

## KEP-3022: min domains in Pod Topology Spread

Pod Topology Spread has the `maxSkew` parameter to define the degree to which Pods may be unevenly distributed. 

> But, there wasn't a way to control the number of domains over which we should spread. 

Reviewer: I would add a sentence along these lines:
kube-scheduler can only operate based on the Nodes that currently exist in a cluster. However, in a cloud environment, the cluster-autoscaler can add or remove Nodes based on demand. As a result, users need to
control the spread over topology domains that might not have any Nodes at a given time.

Some users want to force spreading Pods over a minimum number of domains, and if there aren't enough already present, make the cluster-autoscaler provision them.

Then, we introduced the `minDomains` parameter in the Pod Topology Spread. 
Via `minDomains` parameter, you can define the minimum number of domains. 

For example, there are 3 Nodes with the enough capacity, 
and newly created replicaset has the following `topologySpreadConstraints` in template.

```yaml
topologySpreadConstraints:
- maxSkew: 1
  minDomains: 5 # requires 5 Nodes at least.
  whenUnsatisfiable: DoNotSchedule # minDomains is valid only when DoNotSchedule is used.
  topologyKey: kubernetes.io/hostname 
  labelSelector:
    matchLabels:
        foo: bar
```

This case, 3 Pods will be scheduled to those 3 Nodes,
but other 2 Pods from this replicaset will be unschedulable until more Nodes join the cluster.

The cluster autoscaler provisions new Nodes based on these unschedulable Pods,
and as a result, the replicas are finally spread over 5 Nodes.

## KEP-3094: Take taints/tolerations into consideration when calculating PodTopologySpread skew

> Before this, when we deploy a pod with `podTopologySpread` configured, we'll take all

Reviewer: ```suggestion
Before this enhancement, when you deploy a pod with `podTopologySpread` configured, kube-scheduler would take all
```

affinity nodes(satisfied with pod nodeAffinity and nodeSelector) into consideration

> in filtering and scoring, but a node with pod untolerated taint may also be a candidate

Reviewer: ```suggestion
in filtering and scoring. However, a Node which the incoming Pod doesn't tolerate taint may also be a candidate
```

Reviewer: Although I didn't quite understand the example

Reviewer 2: Polished the description.

because we didn't take care of node taints, which will lead to the pod pending.

To avoid this and make a more fine-gained decision in scheduling, we introduced two new fields in
`TopologySpreadConstraint` to define node inclusion policies including nodeAffinity and nodeTaint.

It mostly looks like:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  # Configure a topology spread constraint
  topologySpreadConstraints:
    - maxSkew: <integer>
      # ...
      nodeAffinityPolicy: [Honor|Ignore]
      nodeTaintsPolicy: [Honor|Ignore]
  # other Pod fields go here
```

**nodeAffinityPolicy** indicates how we'll treat Pod's nodeAffinity/nodeSelector in pod topology spreading.
If `Honor`, we'll filter out nodes not matching nodeAffinity/nodeSelector in calculation.
If `Ignore`, these nodes will be included instead.

For backwards-compatibility, nodeAffinityPolicy is default to `Honor`.

**nodeTaintsPolicy** indicates how we'll treat node taints in pod topology spreading.

> If `Honor`, only tainted nodes for which the incoming pod has a toleration, will be included in calculation.

Reviewer: ```suggestion
If `Honor`, only tainted nodes for which the incoming pod has a toleration, will be included in the calculation of spreading.
```

If `Ignore`, we'll not consider the node taints at all in calculation, so a node with pod untolerated taint
will also be included.

For backwards-compatibility, nodeTaintsPolicy is default to the `Ignore`.

The feature was introduced in v1.25 as alpha level. By default, it was disabled, so if you want to use this feature in v1.25,
you have to enable the feature gate `NodeInclusionPolicyInPodTopologySpread` actively. In the following v1.26, we graduated
this feature to beta and it was enabled by default since.

## KEP-3243: Respect PodTopologySpread after rolling upgrades

> Pod Topology Spread uses the fields `topologyKey` or `labelSelector` to identify the group of pods over which

Reviewer: ```suggestion
Pod Topology Spread uses the field `labelSelector` to identify the group of pods over which
```

Reviewer 3: Done

> spreading will be calculated. But it applies to all pods in a Deployment irrespective of their owning

Reviewer: This is inaccurate. It depends on which label selector you use.

Maybe you can start like this:
When using topology spreading with Deployments, it is common practice to use the labelSelector of the Deployment as the label selector in the topology spread constraints. However, this implies that all pods of a Deployment are part of the spreading calculation, regardless of whether they belong to different revisions.

Reviewer 3: Done

ReplicaSet. As a result, when a new revision is rolled out, spreading will apply across pods from both the
old and new ReplicaSets, and so by the time the new ReplicaSet is completely rolled out and the old one is
rolled back, the actual spreading we are left with may not match expectations because the deleted pods from
the older ReplicaSet will cause skewed distribution for the remaining pods.

In order to solve this problem and to make more accurate decisions in scheduling, we added a new named
`matchLabelKeys` to `topologySpreadConstraints`. `matchLabelKeys` is a list of pod label keys to select

> the pods over which spreading will be calculated. The keys are used to lookup values from the pod labels,

Reviewer: ```suggestion
the pods over which spreading will be calculated. The keys are used to lookup values from the labels of the Pod being scheduled,
```

Reviewer 3: Done

those key-value labels are ANDed with `labelSelector` to select the group of existing pods over
which spreading will be calculated for the incoming pod.

> With `matchLabelKeys`, you don't need to update the `pod.spec` between different revisions.

Reviewer: It's not explained why you would do that. When describing the original problem (two paragraphs above) you can say that an alternative would be to add a different label value every time the deployment changes.

Reviewer 3: OK, yeah. I add this in the paragraphs above.

---
layout: blog
title: "Kubernetes 1.27: More fine-grained pod topology spread policies reached beta"
date: 2023-04-11
slug: fine-grained-pod-topology-spread-features-beta

> evergreen: true

Author: > This represents a commitment by the SIG to keep the content current.

Ah, OK. I'll delete then.

---

**Authors:** [Alex Wang](https://github.com/denkensk)(Shopee), [Kante Yin](https://github.com/kerthcet)(DaoCloud), [Kensei Nakada](https://github.com/sanposhiho)(Mercari)

> In Kubernetes v1.19, [Pod Topology Spread Constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/) went to GA.

Reviewer 4: (nit)
```suggestion
In Kubernetes v1.19, [Pod topology spread constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/)
went to general availability (GA).
```

> It is the feature to control how Pods are spread to each failure-domain (regions, zones, nodes etc).

Reviewer: ```suggestion
It is the feature to control how Pods are spread in the cluster topology or failure domains (regions, zones, nodes etc).
```

> As time passed, we received feedback from users,

Reviewer 4: ```suggestion
As time passed, we - SIG Scheduling - received feedback from users,
```

and, as a result, we're actively working on improving the Topology Spread feature via three KEPs.
All of these features have reached beta in Kubernetes v1.27 and are enabled by default.

This blog post introduces each feature and the use case behind each of them.

## KEP-3022: min domains in Pod Topology Spread

Pod Topology Spread has the `maxSkew` parameter to define the degree to which Pods may be unevenly distributed.

But, there wasn't a way to control the number of domains over which we should spread.
Some users want to force spreading Pods over a minimum number of domains, and if there aren't enough already present, make the cluster-autoscaler provision them.

Then, we introduced the `minDomains` parameter in the Pod Topology Spread.
Via `minDomains` parameter, you can define the minimum number of domains.

For example, assume there are 3 Nodes with the enough capacity,
and a newly created replicaset has the following `topologySpreadConstraints` in template.

```yaml
topologySpreadConstraints:
- maxSkew: 1

>   minDomains: 5 # requires 5 Nodes at least.

Reviewer 4: ```suggestion
  minDomains: 5 # requires 5 Nodes at least (because each Node has a unique hostname)
```

  whenUnsatisfiable: DoNotSchedule # minDomains is valid only when DoNotSchedule is used.
  topologyKey: kubernetes.io/hostname
  labelSelector:
    matchLabels:
        foo: bar
```

In this case, 3 Pods will be scheduled to those 3 Nodes,
but other 2 Pods from this replicaset will be unschedulable until more Nodes join the cluster.

The cluster autoscaler provisions new Nodes based on these unschedulable Pods,
and as a result, the replicas are finally spread over 5 Nodes.

## KEP-3094: Take taints/tolerations into consideration when calculating PodTopologySpread skew

Before this enhancement, when you deploy a pod with `podTopologySpread` configured, kube-scheduler would

> take all inclined nodes(satisfied with pod nodeAffinity and nodeSelector) into consideration

Reviewer: ```suggestion
take the Nodes that satisfy the Pod's nodeAffinity and nodeSelector into consideration
```

in filtering and scoring, but would not care about whether the node taints are tolerated by the incoming pod or not.

> This may lead to a node with untolerated taint best fit the pod in podTopologySpread plugin, and as a result,

Reviewer: ```suggestion
This may lead to a node with untolerated taint as the only candidate for spreading, and as a result,
```

> the pod will stuck in pending for it violates the nodeTaint plugin.

Reviewer: ```suggestion
the pod will stuck in Pending if it doesn't tolerate the taint.
```

 To allow more fine-gained decisions about which Nodes to account for when calculating spreading skew, we introduced

>  two new fields in `TopologySpreadConstraint` to define node inclusion policies including nodeAffinity and nodeTaint.

Reviewer 2: I think so, what's your concern here?

A manifest that applies these policies looks like the following:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  # Configure a topology spread constraint
  topologySpreadConstraints:
    - maxSkew: <integer>
      # ...
      nodeAffinityPolicy: [Honor|Ignore]
      nodeTaintsPolicy: [Honor|Ignore]
  # other Pod fields go here
```

**nodeAffinityPolicy** indicates how we'll treat Pod's nodeAffinity/nodeSelector in pod topology spreading.
If `Honor`, kube-scheduler will filter out nodes not matching nodeAffinity/nodeSelector in the calculation of spreading skew.
If `Ignore`, all nodes will be included, regardless of whether they match the Pod's nodeAffinity/nodeSelector or not.

> For backwards-compatibility, nodeAffinityPolicy defaults to `Honor`.

Reviewer 4: ```suggestion
For backwards compatibility, `nodeAffinityPolicy` defaults to `Honor`.
```

> **nodeTaintsPolicy** indicates how we'll treat node taints in pod topology spreading.

Reviewer 4: ```suggestion
The `nodeTaintsPolicy` field defines how Kubernetes considers node taints for pod topology spreading.
```

If `Honor`, only tainted nodes for which the incoming pod has a toleration, will be included in the calculation of spreading skew.
If `Ignore`, kube-scheduler will not consider the node taints at all in the calculation of spreading skew, so a node with
pod untolerated taint will also be included.

> For backwards-compatibility, nodeTaintsPolicy defaults to the `Ignore`.

Reviewer 4: ```suggestion
For backwards compatibility, `nodeTaintsPolicy` defaults to `Ignore`.
```

> The feature was introduced in v1.25 as alpha level. By default, it was disabled, so if you want to use this feature in v1.25,
> you have to enable the feature gate `NodeInclusionPolicyInPodTopologySpread` actively. In the following v1.26, we graduated
> this feature to beta and it was enabled by default since.

Reviewer 4: ```suggestion
The feature was introduced in v1.25 as alpha. By default, it was disabled, so if you want to use this feature in v1.25,
you had to explictly enable the feature gate `NodeInclusionPolicyInPodTopologySpread`. In the following v1.26
release, that associated feature graduated to beta and is enabled by default.
```

> ## KEP-3243: Respect PodTopologySpread after rolling upgrades

Reviewer 4: ```suggestion
## KEP-3243: Respect Pod topology spread after rolling upgrades
```
This isn't the exact title, but it's close enough

Pod Topology Spread uses the field `labelSelector` to identify the group of pods over which
spreading will be calculated. When using topology spreading with Deployments, it is common
practice to use the `labelSelector` of the Deployment as the `labelSelector` in the topology
spread constraints. However, this implies that all pods of a Deployment are part of the spreading
calculation, regardless of whether they belong to different revisions. As a result, when a new revision
is rolled out, spreading will apply across pods from both the old and new ReplicaSets, and so by the
time the new ReplicaSet is completely rolled out and the old one is rolled back, the actual spreading
we are left with may not match expectations because the deleted pods from the older ReplicaSet will cause
skewed distribution for the remaining pods. To avoid this problem, in the past users needed to add a
revision label to Deployment and update it manually at each rolling upgrade (both the label on the

> podTemplate and the `labelSelector` in the `topologySpreadConstraints`).

Reviewer 4: ```suggestion
Pod template and the `labelSelector` in the `topologySpreadConstraints`).
```

> To solve this problem with a simpler API, we added a new field named

Reviewer 4: ```suggestion
To solve this problem with a simpler API, Kubernetes v1.25 introduced a new field named
```

> To solve this problem once and for all, and to make more accurate decisions in scheduling, we added a new named

Reviewer: ```suggestion
To solve this problem with a simpler API, we added a new field named
```

the pods over which spreading will be calculated. The keys are used to lookup values from the labels of
the Pod being scheduled, those key-value labels are ANDed with `labelSelector` to select the group of
existing pods over which spreading will be calculated for the incoming pod.

With `matchLabelKeys`, you don't need to update the `pod.spec` between different revisions.
The controller or operator managing rollouts just needs to set different values to the same label key for different revisions.
The scheduler will assume the values automatically based on `matchLabelKeys`.
For example, if you are configuring a Deployment, you can use the label keyed with
[pod-template-hash](/docs/concepts/workloads/controllers/deployment/#pod-template-hash-label),
which is added automatically by the Deployment controller, to distinguish between different
revisions in a single Deployment.

```yaml
topologySpreadConstraints:
    - maxSkew: 1
      topologyKey: kubernetes.io/hostname
      whenUnsatisfiable: DoNotSchedule
      labelSelector:
        matchLabels:
          app: foo
      matchLabelKeys:
        - pod-template-hash
```

## Getting involved

> These features are managed by the [SIG/Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling).

Reviewer 4: ```suggestion
These features are managed by Kubernetes [SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling).
```

Please join us and share your feedback. We look forward to hearing from you!

## How can I learn more?

> - [Pod Topology Spread Constraints | Kubernetes](/docs/concepts/scheduling-eviction/topology-spread-constraints/)

Reviewer 4: ```suggestion
- [Pod Topology Spread Constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/) in the Kubernetes documentation
```

> - [Pod Topology Spread Constraints | Kubernetes](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#container-resource-metrics)

Reviewer 2: @sanposhiho is this by mistake, I suppose it should be https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/.

Author: Oops, nice catch 😅

---
layout: blog

> title: "TBD" // TODO: have a cool title.

Author: @denkensk @kerthcet any ideas? 🙂

Reviewer 3: How about <Introducing the advanced pod topology spread policy for Kubernetes>?

Reviewer 2: or `More fine-gained pod topology spread policies in Kubernetes scheduling`.

Author: I like the word "fine-gained". 
Also, I'd like to mention that these features are coming to beta in v1.27 like [many other blog posts did](https://kubernetes.io/blog/page/3/).

So... like `Kubernetes 1.27: More fine-gained pod topology spread policies reached beta`?

Reviewer 3: `Kubernetes 1.27: More fine-grained pod topology spread policies reached beta` 
sgtm.

date: 2023-04-11

> slug: topology-spread-new-features

Reviewer 2: `fine-gained-pod-topology-spread-features-beta` ?

evergreen: true
---

> **Authors:** [Alex Wang](https://github.com/denkensk)(), [Kante Yin](https://github.com/kerthcet)(), [Kensei Nakada](https://github.com/sanposhiho)(Mercari)

Author: (Just an alphabetical order)

Author: @denkensk @kerthcet 
I'm not sure where you both are working for. Give me your companies, please.

Reviewer 2: DaoCloud 😄

Author: Thanks! 👍
